src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/EditingTargetInlineChunkExecution.java

```diff
@@ -0,0 +1,184 @@
+   {
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      
+      display_ = display;
+      docId_ = docId;
+      outputs_ = new HashMap<String, ChunkInlineOutput>();
+   }
+   
+   public void execute(Range range)
+   {
+      // synthesize an identifier for this chunk execution
+      final String chunkId = "i" + StringUtil.makeRandomId(12);
+      
+      // check to see if we're already showing a panel for this range; if we
+      // are, remove it to make way for the new one
+      for (ChunkInlineOutput output: outputs_.values())
+      {
+         if (output.range().isEqualTo(range))
+         {
+            if (output.state() == ChunkInlineOutput.State.Finished)
+            {
+               // remove old, completed output for this input
+               output.hide();
+               outputs_.remove(output.chunkId());
+            }
+            else
+            {
+               // we already have an output panel for this input, and it's
+               // either waiting to execute or currently executing. ignore this
+               // request to re-execute the range, as it's likely to be an
+               // unintended duplicate.
+               return;
+            }
+         }
+      }
+      
+      // create dummy scope for execution
+      Scope scope = Scope.createRScopeNode(
+            chunkId,
+            range.getStart(),
+            range.getEnd(),
+            Scope.SCOPE_TYPE_CHUNK);
+      
+      // create popup panel to host output
+      final ChunkInlineOutput output = new ChunkInlineOutput(chunkId, range);
+      output.addCloseHandler(new CloseHandler<PopupPanel>()
+      {
+         @Override
+         public void onClose(CloseEvent<PopupPanel> event)
+         {
+            outputs_.remove(chunkId);
+         }
+      });
+
+      // render offscreen until complete
+      output.setPopupPosition(-100000, -100000);
+      output.show();
+      outputs_.put(chunkId, output);
+
+      SendToChunkConsoleEvent event =
+            new SendToChunkConsoleEvent(docId_, scope, range, 
+                  NotebookQueueUnit.EXEC_SCOPE_INLINE);
+      events_.fireEvent(event);
+   }
+   
+   // Handlers ----
+   
+   @Override
+   public void onConsoleWriteError(ConsoleWriteErrorEvent event)
+   {
+      if (outputs_.containsKey(event.getConsole()))
+        outputs_.get(event.getConsole()).onConsoleWriteError(event);
+   }
+
+   @Override
+   public void onConsoleWriteOutput(ConsoleWriteOutputEvent event)
+   {
+      if (outputs_.containsKey(event.getConsole()))
+        outputs_.get(event.getConsole()).onConsoleWriteOutput(event);
+   }
+
+   @Override
+   public void onChunkExecStateChanged(ChunkExecStateChangedEvent event)
+   {
+      // ignore if not targeted at one of our chunks
+      if (event.getDocId() != docId_ || 
+          !outputs_.containsKey(event.getChunkId()))
+         return;
+      
+      ChunkInlineOutput output = outputs_.get(event.getChunkId());
```

Comment:
It'll always succeed (we bail out if the hashmap doesn't contain the chunk ID a few lines above).